exareme2/controller/services/exaflow/execution_engine.py

```diff
@@ -0,0 +1,42 @@
+from typing import List
+
+from exareme2.controller import logger as ctrl_logger
+from exareme2.controller.services.exaflow.tasks_handler import TasksHandler
+
+
+class AlgorithmExecutionEngine:
+    """
+    The AlgorithmExecutionEngine is the class used by the algorithms to communicate with
+    the workers of the system. An AlgorithmExecutionEngine object is passed to all algorithms
+    """
+
+    def __init__(
+        self,
+        request_id: str,
+        context_id: str,
+        csv_paths_per_worker_id: dict,
+        tasks_handlers: List[TasksHandler],
+    ):
+        self._logger = ctrl_logger.get_request_logger(request_id=request_id)
+        self._context_id = context_id
+        self._tasks_handlers = tasks_handlers
+        self._csv_paths_per_worker_id = csv_paths_per_worker_id
+
+    def run_algorithm_step(
+        self,
+        func,
+        positional_args,
+    ) -> List[dict]:
+        # Queue the udf on all local workers
+        tasks = {}
+        for task_handler in self._tasks_handlers:
+            current_positional_args = positional_args
+            current_positional_args["csv_paths"] = self._csv_paths_per_worker_id[
+                task_handler.worker_id
+            ]
+            tasks[task_handler] = task_handler.queue_udf(
+                udf_name=func,
+                params=current_positional_args,
+            )
+
+        return [worker.get_udf_result(task) for worker, task in tasks.items()]
```

Comment:
This can be simplified by keeping the results in a list and then just calling their `get` method.